elements/lisk-transactions/src/15_proof_of_misbehavior_transaction.ts

```diff
@@ -0,0 +1,442 @@
+				),
+			);
+		}
+
+		if (delegateAccount.delegate.isBanned) {
+			errors.push(
+				new TransactionError(
+					'Cannot apply proof-of-misbehavior. Delegate is banned.',
+					this.id,
+					'.asset.header1.generatorPublicKey',
+					this.asset.header1.generatorPublicKey,
+				),
+			);
+		}
+
+		if (
+			isPunished(
+				delegateAccount,
+				delegateAccount,
+				store.chain.lastBlockHeader.height,
+			)
+		) {
+			errors.push(
+				new TransactionError(
+					'Cannot apply proof-of-misbehavior. Delegate is already punished. ',
+					this.id,
+					'.asset.header1.generatorPublicKey',
+					this.asset.header1.generatorPublicKey,
+				),
+			);
+		}
+
+		/* 
+			Check block signatures validity 
+		*/
+
+		const blockHeader1Bytes = Buffer.concat([
+			Buffer.from(networkIdentifier, 'hex'),
+			getBlockBytes(this.asset.header1),
+		]);
+		const blockHeader2Bytes = Buffer.concat([
+			Buffer.from(networkIdentifier, 'hex'),
+			getBlockBytes(this.asset.header2),
+		]);
+
+		const { valid: validHeader1Signature } = validateSignature(
+			this.asset.header1.generatorPublicKey,
+			this.asset.header1.blockSignature,
+			blockHeader1Bytes,
+		);
+
+		if (!validHeader1Signature) {
+			errors.push(
+				new TransactionError(
+					'Invalid block signature for header 1.',
+					this.id,
+					'.asset.header1.blockSignature',
+					this.asset.header1.blockSignature,
+				),
+			);
+		}
+		const { valid: validHeader2Signature } = validateSignature(
+			this.asset.header2.generatorPublicKey,
+			this.asset.header2.blockSignature,
+			blockHeader2Bytes,
+		);
+
+		if (!validHeader2Signature) {
+			errors.push(
+				new TransactionError(
+					'Invalid block signature for header 2.',
+					this.id,
+					'.asset.header2.blockSignature',
+					this.asset.header2.blockSignature,
+				),
+			);
+		}
+
+		/*
+			Update sender account
+		*/
+		const reward =
+			store.chain.lastBlockReward > delegateAccount.balance
+				? delegateAccount.balance
+				: store.chain.lastBlockReward;
+
+		senderAccount.balance += reward;
+		// We store the correct reward value in the asset at apply time to allow for undoing the transaction at a later point in time
+		this.asset.reward = reward;
+
+		store.account.set(senderAccount.address, senderAccount);
+
+		/*
+			Update delegate account
+		*/
+		delegateAccount.delegate.pomHeights.push(currentHeight);
+
+		// tslint:disable-next-line no-magic-numbers
+		if (delegateAccount.delegate.pomHeights.length > 4) {
+			delegateAccount.delegate.isBanned = true;
+		}
+		delegateAccount.balance -= reward;
+		store.account.set(delegateAccount.address, delegateAccount);
+
+		return errors;
+	}
+
+	protected async undoAsset(
+		store: StateStore,
+	): Promise<ReadonlyArray<TransactionError>> {
+		const currentHeight = store.chain.lastBlockHeader.height + 1;
+		const delegateId = getAddressFromPublicKey(
+			this.asset.header1.generatorPublicKey,
+		);
+		const delegateAccount = await store.account.get(delegateId);
```

Comment:
same as above, there is a case delegate account and sender account is the same.
so get(delegateId) should happen after set(senderAccount)